File: src/api/http/git/streaming.ts
/**
 * Git Streaming Operations
 *
 * SSE-based streaming for git push, pull, fetch operations.
 * Provides real-time output from git commands via Server-Sent Events.
 */
import { createSSEStream } from "@src/api/realtime/sseStream";

import { gitRepoUrl } from "./client";

/**
 * Git error types detected by the backend
 * These map to specific dialogs in the frontend
 */
export type GitErrorType =
  | "none"
  | "non_fast_forward" // Push rejected - remote has changes
  | "protected_branch" // Target branch is protected
  | "authentication_failed" // Auth failed
  | "remote_branch_deleted" // Remote branch was deleted
  | "uncommitted_changes" // Local changes would be overwritten
  | "network_error" // Network/connection error
  | "merge_conflicts" // Merge conflicts
  | "permission_denied" // Permission denied
  | "unknown"; // Unknown error

export interface GitStreamResult {
  success: boolean;
  errorType: GitErrorType;
}

export interface GitStreamCallbacks {
  onStart?: () => void;
  onOutput?: (line: string, stream: "stdout" | "stderr") => void;
  /** Called when operation completes with success status and error type */
  onComplete?: (success: boolean, errorType?: GitErrorType) => void;
  onError?: (error: string, errorType?: GitErrorType) => void;
}

/**
 * Stream git push output in real-time
 *
 * @example
 * ```typescript
 * const cleanup = await gitPushStream(
 *   {
 *     repo_id: "my-repo",
 *     repo_path: "/path/to/repo",
 *     remote: "origin",
 *     set_upstream: true,
 *   },
 *   {
 *     onOutput: (line, stream) => *
 * // Cancel if needed
 * cleanup();
 * ```
 */
export async function gitPushStream(
  params: {
    repo_id: string;
    repo_path: string;
    remote?: string;
    branch?: string;
    set_upstream?: boolean;
    force?: boolean;
  },
  callbacks: GitStreamCallbacks
): Promise<() => void> {
  const queryParams = new URLSearchParams({
    path: params.repo_path,
  });

  if (params.remote) queryParams.append("remote", params.remote);
  if (params.branch) queryParams.append("branch", params.branch);
  if (params.set_upstream) queryParams.append("set_upstream", "true");
  if (params.force) queryParams.append("force", "true");

  const url = `${gitRepoUrl(params.repo_id)}/push/stream?${queryParams.toString()}`;

  return createSSEStream({
    url,
    onStart: () => {
      callbacks.onStart?.();
    },
    onOutput: (data) => {
      callbacks.onOutput?.(data.line, data.stream as "stdout" | "stderr");
    },
    onEnd: (data) => {
      const errorType = (data.error_type as GitErrorType) || "none";
      callbacks.onComplete?.(data.success, errorType);
    },
    onError: (error, data) => {
      const errorType = (data?.error_type as GitErrorType) || "unknown";
      callbacks.onError?.(error, errorType);
    },
  });
}

/**
 * Stream git pull output in real-time
 *
 * @example
 * ```typescript
 * const cleanup = await gitPullStream(
 *   {
 *     repo_id: "my-repo",
 *     repo_path: "/path/to/repo",
 *     remote: "origin",
 *   },
 *   {
 *     onOutput: (line) => * ```
 */
export async function gitPullStream(
  params: {
    repo_id: string;
    repo_path: string;
    remote?: string;
    branch?: string;
    strategy?: string;
  },
  callbacks: GitStreamCallbacks
): Promise<() => void> {
  const queryParams = new URLSearchParams({
    path: params.repo_path,
  });

  if (params.remote) queryParams.append("remote", params.remote);
  if (params.branch) queryParams.append("branch", params.branch);
  if (params.strategy) queryParams.append("strategy", params.strategy);

  const url = `${gitRepoUrl(params.repo_id)}/pull/stream?${queryParams.toString()}`;

  return createSSEStream({
    url,
    onStart: () => callbacks.onStart?.(),
    onOutput: (data) =>
      callbacks.onOutput?.(data.line, data.stream as "stdout" | "stderr"),
    onEnd: (data) => {
      const errorType = (data.error_type as GitErrorType) || "none";
      callbacks.onComplete?.(data.success, errorType);
    },
    onError: (error, data) => {
      const errorType = (data?.error_type as GitErrorType) || "unknown";
      callbacks.onError?.(error, errorType);
    },
  });
}

/**
 * Stream git fetch output in real-time
 *
 * @example
 * ```typescript
 * const cleanup = await gitFetchStream(
 *   {
 *     repo_id: "my-repo",
 *     repo_path: "/path/to/repo",
 *     remote: "origin",
 *   },
 *   {
 *     onOutput: (line) => * ```
 */
export async function gitFetchStream(
  params: {
    repo_id: string;
    repo_path: string;
    remote?: string;
    prune?: boolean;
  },
  callbacks: GitStreamCallbacks
): Promise<() => void> {
  const queryParams = new URLSearchParams({
    path: params.repo_path,
  });

  if (params.remote) queryParams.append("remote", params.remote);
  if (params.prune !== undefined)
    queryParams.append("prune", params.prune.toString());

  const url = `${gitRepoUrl(params.repo_id)}/fetch/stream?${queryParams.toString()}`;

  return createSSEStream({
    url,
    onStart: () => callbacks.onStart?.(),
    onOutput: (data) =>
      callbacks.onOutput?.(data.line, data.stream as "stdout" | "stderr"),
    onEnd: (data) => {
      const errorType = (data.error_type as GitErrorType) || "none";
      callbacks.onComplete?.(data.success, errorType);
    },
    onError: (error, data) => {
      const errorType = (data?.error_type as GitErrorType) || "unknown";
      callbacks.onError?.(error, errorType);
    },
  });
}

/**
 * Stream git commit output in real-time
 */
export async function gitCommitStream(
  params: {
    repo_id: string;
    repo_path: string;
    message: string;
  },
  callbacks: GitStreamCallbacks
): Promise<() => void> {
  const queryParams = new URLSearchParams({
    path: params.repo_path,
    message: params.message,
  });

  const url = `${gitRepoUrl(params.repo_id)}/commit/stream?${queryParams.toString()}`;

  return createSSEStream({
    url,
    onStart: () => callbacks.onStart?.(),
    onOutput: (data) =>
      callbacks.onOutput?.(data.line, data.stream as "stdout" | "stderr"),
    onEnd: (data) => {
      const errorType = (data.error_type as GitErrorType) || "none";
      callbacks.onComplete?.(data.success, errorType);
    },
    onError: (error, data) => {
      const errorType = (data?.error_type as GitErrorType) || "unknown";
      callbacks.onError?.(error, errorType);
    },
  });
}

/**
 * Stream git add (stage) output in real-time
 */
export async function gitStageStream(
  params: {
    repo_id: string;
    repo_path: string;
    files: string[];
  },
  callbacks: GitStreamCallbacks
): Promise<() => void> {
  const queryParams = new URLSearchParams({
    path: params.repo_path,
    files: JSON.stringify(params.files),
  });

  const url = `${gitRepoUrl(params.repo_id)}/stage/stream?${queryParams.toString()}`;

  return createSSEStream({
    url,
    onStart: () => callbacks.onStart?.(),
    onOutput: (data) =>
      callbacks.onOutput?.(data.line, data.stream as "stdout" | "stderr"),
    onEnd: (data) => {
      const errorType = (data.error_type as GitErrorType) || "none";
      callbacks.onComplete?.(data.success, errorType);
    },
    onError: (error, data) => {
      const errorType = (data?.error_type as GitErrorType) || "unknown";
      callbacks.onError?.(error, errorType);
    },
  });
}

File: src-tauri/crates/git-api/src/commands/streaming.rs
/**
 * Streaming Git Operations with Server-Sent Events (SSE)
 *
 * Provides real-time output streaming for long-running git operations.
 * Uses Server-Sent Events (SSE) for efficient one-way server→client streaming.
 *
 * NOTE: Uses pre_exec on Unix to close inherited file descriptors (3-1024)
 * to prevent "Bad file descriptor" errors from WebView FD inheritance.
 */
#[cfg(test)]
#[path = "tests/streaming_tests.rs"]
mod tests;

use axum::{
    extract::{Path, Query},
    response::sse::{Event, KeepAlive},
    response::{IntoResponse, Response, Sse},
};
use futures::stream::Stream;
use serde::Deserialize;
use std::path::PathBuf;
use std::pin::Pin;
use std::process::Stdio;
use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::process::Command;

use git::tokio_git_command;

type GitEventStream = Pin<Box<dyn Stream<Item = Result<Event, std::convert::Infallible>> + Send>>;

// ============================================
// Error Type Detection
// ============================================

/// Detect error type from combined output (for streaming responses)
pub(crate) fn detect_error_type_from_output(output: &str, operation: &str) -> &'static str {
    let lower = output.to_lowercase();

    match operation {
        "push" => {
            // Non-fast-forward (remote has changes we don't have)
            if lower.contains("non-fast-forward")
                || lower.contains("fetch first")
                || lower.contains("updates were rejected")
                || lower.contains("failed to push some refs")
            {
                return "non_fast_forward";
            }

            // Protected branch
            if lower.contains("protected branch")
                || lower.contains("branch is protected")
                || lower.contains("cannot push to")
                || lower.contains("pre-receive hook declined")
                || lower.contains("remote rejected")
            {
                return "protected_branch";
            }
        }
        "pull" => {
            // Uncommitted changes would be overwritten
            if lower.contains("would be overwritten")
                || lower.contains("your local changes")
                || lower.contains("uncommitted changes")
                || lower.contains("please commit your changes or stash them")
            {
                return "uncommitted_changes";
            }

            // Merge conflicts
            if lower.contains("conflict") || lower.contains("automatic merge failed") {
                return "merge_conflicts";
            }
        }
        "fetch" => {
            // Check for deleted branches
            if lower.contains("[deleted]") {
                return "remote_branch_deleted";
            }
        }
        _ => {}
    }

    // Common errors across all operations
    if lower.contains("authentication failed")
        || lower.contains("invalid credentials")
        || lower.contains("invalid username or password")
        || lower.contains("invalid username or token")
        || lower.contains("bad credentials")
        || lower.contains("http basic: access denied")
        || lower.contains("could not read username")
        || lower.contains("unable to get password from user")
        || lower.contains("permission denied (publickey)")
        || lower.contains("repository not found")
        || lower.contains("saml")
        || lower.contains("sso")
        || lower.contains("password authentication was removed")
        || lower.contains("requested url returned error: 403")
    {
        return "authentication_failed";
    }

    if lower.contains("could not resolve host")
        || lower.contains("connection refused")
        || lower.contains("network is unreachable")
        || lower.contains("unable to access")
        || lower.contains("connection timed out")
    {
        return "network_error";
    }

    "unknown"
}

// ============================================
// Query Parameters
// ============================================

#[derive(Deserialize)]
pub struct PushStreamQuery {
    pub path: String,
    #[serde(default)]
    pub remote: Option<String>,
    #[serde(default)]
    pub branch: Option<String>,
    #[serde(default)]
    pub set_upstream: Option<bool>,
    #[serde(default)]
    pub force: Option<bool>,
}

#[derive(Deserialize)]
pub struct PullStreamQuery {
    pub path: String,
    #[serde(default)]
    pub remote: Option<String>,
    #[serde(default)]
    pub branch: Option<String>,
    /// Pull strategy: "merge" (default), "rebase", or "ff-only"
    #[serde(default)]
    pub strategy: Option<String>,
}

#[derive(Deserialize)]
pub struct FetchStreamQuery {
    pub path: String,
    #[serde(default)]
    pub remote: Option<String>,
    #[serde(default)]
    pub prune: Option<bool>,
}

#[derive(Deserialize)]
pub struct CommitStreamQuery {
    pub path: String,
    pub message: String,
}

#[derive(Deserialize)]
pub struct StageStreamQuery {
    pub path: String,
    pub files: String, // JSON array of files
}

/// Check if an error is a transient system error that can be retried
pub(crate) fn is_transient_error(error_msg: &str) -> bool {
    error_msg.contains("Bad file descriptor")
        || error_msg.contains("Resource temporarily unavailable")
        || error_msg.contains("os error 9")
        || error_msg.contains("Too many open files")
        || error_msg.contains("os error 24")
}

/// Configure command with pre_exec to close inherited file descriptors on Unix
/// This prevents "Bad file descriptor" errors from WebView FD inheritance
#[cfg(unix)]
fn configure_command_for_fd_safety(cmd: &mut Command) {
    // SAFETY: We only close file descriptors, which is safe
    unsafe {
        cmd.pre_exec(|| {
            // Close file descriptors 3-1024 to avoid inheriting bad FDs from WebView
            for fd in 3..1024 {
                libc::close(fd);
            }
            Ok(())
        });
    }
}

#[cfg(not(unix))]
fn configure_command_for_fd_safety(_cmd: &mut Command) {
    // No-op on non-Unix systems
}

/// Helper function to stream git command output with retry logic for transient errors
/// Retries up to 3 times on "Bad file descriptor" and similar transient errors
/// The `operation` parameter is used for error type detection ("push", "pull", "fetch", etc.)
async fn stream_git_command(
    mut cmd: Command,
    command_str: String,
    operation: &'static str,
) -> GitEventStream {
    // Configure command for file descriptor safety
    configure_command_for_fd_safety(&mut cmd);

    let stream = async_stream::stream! {
        yield Ok(Event::default()
            .event("start")
            .data(format!("{{\"command\":\"{}\"}}", command_str)));

        const MAX_RETRIES: u32 = 3;
        let mut attempt = 0;

        loop {
            attempt += 1;

            let mut child = match cmd.spawn() {
                Ok(child) => child,
                Err(e) => {
                    let error_str = e.to_string();

                    // Check if this is a transient error that can be retried
                    if is_transient_error(&error_str) && attempt < MAX_RETRIES {
                        // Wait briefly before retrying (exponential backoff)
                        tokio::time::sleep(tokio::time::Duration::from_millis(100 * attempt as u64)).await;
                        continue;
                    }

                    let error_msg = error_str.replace("\"", "\\\"");
                    yield Ok(Event::default()
                        .event("error")
                        .data(format!("{{\"error\":\"{}\",\"error_type\":\"unknown\"}}", error_msg)));
                    return;
                }
            };

            let stdout = child.stdout.take();
            let stderr = child.stderr.take();

            let mut all_lines: Vec<(String, String)> = Vec::new();

            if let Some(stdout) = stdout {
                let reader = BufReader::new(stdout);
                if let Ok(lines) = tokio::spawn(async move {
                    let mut lines_reader = reader.lines();
                    let mut output = Vec::new();
                    while let Ok(Some(line)) = lines_reader.next_line().await {
                        output.push(("stdout".to_string(), line));
                    }
                    output
                }).await {
                    all_lines.extend(lines);
                }
            }

            if let Some(stderr) = stderr {
                let reader = BufReader::new(stderr);
                if let Ok(lines) = tokio::spawn(async move {
                    let mut lines_reader = reader.lines();
                    let mut output = Vec::new();
                    while let Ok(Some(line)) = lines_reader.next_line().await {
                        output.push(("stderr".to_string(), line));
                    }
                    output
                }).await {
                    all_lines.extend(lines);
                }
            }

            // Check if any output line indicates a transient error
            let has_transient_error = all_lines.iter().any(|(_, line)| is_transient_error(line));

            if has_transient_error && attempt < MAX_RETRIES {
                // Wait briefly before retrying
                tokio::time::sleep(tokio::time::Duration::from_millis(100 * attempt as u64)).await;
                continue;
            }

            // Collect all output for error type detection
            let combined_output: String = all_lines.iter()
                .map(|(_, line)| line.as_str())
                .collect::<Vec<_>>()
                .join("\n");

            for (stream, line) in &all_lines {
                let escaped_line = line.replace("\"", "\\\"").replace("\n", "\\n");
                yield Ok(Event::default()
                    .event("output")
                    .data(format!("{{\"stream\":\"{}\",\"line\":\"{}\"}}", stream, escaped_line)));
            }

            match child.wait().await {
                Ok(status) => {
                    let error_type = if status.success() {
                        "none"
                    } else {
                        detect_error_type_from_output(&combined_output, operation)
                    };

                    yield Ok(Event::default()
                        .event("end")
                        .data(format!("{{\"success\":{},\"error_type\":\"{}\"}}", status.success(), error_type)));
                }
                Err(e) => {
                    let error_msg = e.to_string().replace("\"", "\\\"");
                    let error_type = detect_error_type_from_output(&error_msg, operation);
                    yield Ok(Event::default()
                        .event("error")
                        .data(format!("{{\"error\":\"{}\",\"error_type\":\"{}\"}}", error_msg, error_type)));
                }
            }

            // Successfully completed, exit the retry loop
            break;
        }
    };

    Box::pin(stream)
}

fn sse_response(stream: GitEventStream) -> Response {
    Sse::new(stream)
        .keep_alive(KeepAlive::default())
        .into_response()
}

fn git_resolution_error_response(error: String) -> Response {
    let escaped = error.replace('"', "\\\"");
    let stream = futures::stream::once(async move {
        Ok(Event::default().event("error").data(format!(
            "{{\"error\":\"{}\",\"error_type\":\"git_unavailable\"}}",
            escaped
        )))
    });

    sse_response(Box::pin(stream) as GitEventStream)
}

// ============================================
// SSE Stream Handlers
// ============================================

/// Stream git push output via Server-Sent Events
pub async fn push_stream(
    Path(_repo_id): Path<String>,
    Query(query): Query<PushStreamQuery>,
) -> Response {
    let repo_path = PathBuf::from(&query.path);
    let remote = query.remote.unwrap_or_else(|| "origin".to_string());
    let set_upstream = query.set_upstream.unwrap_or(false);
    let force = query.force.unwrap_or(false);

    let mut cmd = match tokio_git_command() {
        Ok(command) => command,
        Err(err) => return git_resolution_error_response(err),
    };
    cmd.args(["-c", "credential.interactive=false", "-c", "core.askPass="])
        .arg("push");

    if set_upstream {
        cmd.arg("-u");
    }
    if force {
        cmd.arg("--force");
    }

    cmd.arg(&remote);
    if let Some(branch) = query.branch {
        cmd.arg(&branch);
    }

    cmd.current_dir(&repo_path)
        .env("GIT_TERMINAL_PROMPT", "0")
        .env("GIT_ASKPASS", "")
        .env("SSH_ASKPASS", "")
        .env("GCM_INTERACTIVE", "Never")
        .env("GCM_MODAL_PROMPT", "0")
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());

    let command_str = format!("git push {}", remote);
    let stream = stream_git_command(cmd, command_str, "push").await;

    sse_response(stream)
}

/// Stream git pull output via Server-Sent Events
pub async fn pull_stream(
    Path(_repo_id): Path<String>,
    Query(query): Query<PullStreamQuery>,
) -> Response {
    let repo_path = PathBuf::from(&query.path);
    let remote = query.remote.unwrap_or_else(|| "origin".to_string());

    let mut cmd = match tokio_git_command() {
        Ok(command) => command,
        Err(err) => return git_resolution_error_response(err),
    };
    cmd.args(["-c", "credential.interactive=false", "-c", "core.askPass="])
        .arg("pull");

    let strategy_flag = match query.strategy.as_deref() {
        Some("rebase") => {
            cmd.arg("--rebase");
            " --rebase"
        }
        Some("ff-only") => {
            cmd.arg("--ff-only");
            " --ff-only"
        }
        _ => {
            cmd.arg("--no-rebase");
            " --no-rebase"
        }
    };

    cmd.arg(&remote);

    if let Some(branch) = query.branch {
        cmd.arg(&branch);
    }

    cmd.current_dir(&repo_path)
        .env("GIT_TERMINAL_PROMPT", "0")
        .env("GIT_ASKPASS", "")
        .env("SSH_ASKPASS", "")
        .env("GCM_INTERACTIVE", "Never")
        .env("GCM_MODAL_PROMPT", "0")
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());

    let command_str = format!("git pull{} {}", strategy_flag, remote);
    let stream = stream_git_command(cmd, command_str, "pull").await;

    sse_response(stream)
}

/// Stream git fetch output via Server-Sent Events
pub async fn fetch_stream(
    Path(_repo_id): Path<String>,
    Query(query): Query<FetchStreamQuery>,
) -> Response {
    let repo_path = PathBuf::from(&query.path);
    let remote = query.remote.unwrap_or_else(|| "origin".to_string());
    let prune = query.prune.unwrap_or(true);

    let mut cmd = match tokio_git_command() {
        Ok(command) => command,
        Err(err) => return git_resolution_error_response(err),
    };
    cmd.args(["-c", "credential.interactive=false", "-c", "core.askPass="])
        .arg("fetch")
        .arg(&remote);

    if prune {
        cmd.arg("--prune");
    }

    cmd.current_dir(&repo_path)
        .env("GIT_TERMINAL_PROMPT", "0")
        .env("GIT_ASKPASS", "")
        .env("SSH_ASKPASS", "")
        .env("GCM_INTERACTIVE", "Never")
        .env("GCM_MODAL_PROMPT", "0")
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());

    let command_str = format!("git fetch {}", remote);
    let stream = stream_git_command(cmd, command_str, "fetch").await;

    sse_response(stream)
}

/// Stream git commit output via Server-Sent Events
pub async fn commit_stream(
    Path(_repo_id): Path<String>,
    Query(query): Query<CommitStreamQuery>,
) -> Response {
    let repo_path = PathBuf::from(&query.path);
    let message = query.message;

    let mut cmd = match tokio_git_command() {
        Ok(command) => command,
        Err(err) => return git_resolution_error_response(err),
    };
    cmd.arg("commit")
        .arg("-m")
        .arg(&message)
        .current_dir(&repo_path)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());

    let command_str = format!("git commit -m \"{}\"", message.replace("\"", "\\\""));
    let stream = stream_git_command(cmd, command_str, "commit").await;

    sse_response(stream)
}

/// Stream git add (stage) output via Server-Sent Events
pub async fn stage_stream(
    Path(_repo_id): Path<String>,
    Query(query): Query<StageStreamQuery>,
) -> Response {
    let repo_path = PathBuf::from(&query.path);
    // The `files` query param is a JSON-encoded array. A silent
    // empty fallback would make `git add` run as a no-op, with
    // the user seeing no error and no effect. Warn so a malformed
    // request from the frontend is visible in logs while still
    // degrading to a no-op (we won't error the SSE stream for a
    // request shape we can't parse).
    let files: Vec<String> = match serde_json::from_str(&query.files) {
        Ok(f) => f,
        Err(err) => {
            tracing::warn!(
                error = %err,
                files_param = %query.files,
                "git::stage_stream: files query param is not a JSON array; running git add as no-op"
            );
            Vec::new()
        }
    };

    let mut cmd = match tokio_git_command() {
        Ok(command) => command,
        Err(err) => return git_resolution_error_response(err),
    };
    cmd.arg("add");
    for file in &files {
        cmd.arg(file);
    }
    cmd.current_dir(&repo_path)
        .env("GIT_TERMINAL_PROMPT", "0")
        .env("GIT_ASKPASS", "")
        .env("SSH_ASKPASS", "")
        .env("GCM_INTERACTIVE", "Never")
        .env("GCM_MODAL_PROMPT", "0")
        .stdin(Stdio::null())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());

    let command_str = if files.is_empty() {
        "git add .".to_string()
    } else {
        format!("git add {}", files.join(" "))
    };

    let stream = stream_git_command(cmd, command_str, "stage").await;

    sse_response(stream)
}
